### PR TITLE
FA-26, getting a bitmap from users Gallery

### PR DESCRIPTION
**Pull request by HamdanSCode**

### CODE REVIEW

File: app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt [hunk 1]
```diff
@@ -37,10 +40,30 @@ class DrawingFieldView @JvmOverloads constructor(
...
+     * used for setting our previous bitmap from rotation, no expectation of having drawn
+     */
     fun setBitmap(bitmap: Bitmap) {
         canvasBitmap = bitmap
     }
 
+    /**
+     * sets the bitmap from the image obtained from the gallery.
+     */
+    fun getBitmapFromUriString(uri: Uri, context: Context) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt]: The ticket talked about making a method to get a bitmap from file, not about getting it in one specific View. Make this a util function in its own object file and have it return a Bitmap

Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt]: okily dokily

File: app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt [hunk 2]
```diff
@@ -37,10 +37,24 @@ class DrawingFieldView @JvmOverloads constructor(
...
     fun setBitmap(bitmap: Bitmap) {
         canvasBitmap = bitmap
     }
 
+    /**
+     * sets the bitmap from some image,currently resizes the bitmap
+     */
+    fun setBitmapFromImageBitmap(imageBitmap: Bitmap, resize: Boolean = true) {
+        if (resize) {
+            canvasBitmap = Bitmap.createScaledBitmap(imageBitmap, width, height, true)
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt]: Making a resize bitmap on the main thread is a no no. An imageview has many properties on how to scale/crop on its own. Use one of those instead

File: app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+     * launches with the purpose of obtaining an image, will open the users gallery
+     */
+    fun launchGalleryImageGetter(contentLauncher: ActivityResultLauncher<String>) {
+        contentLauncher.launch("image/*")
+    }
+
+    /**
+     * function that takes a Uri and gets a bitmap from it from the gallery.
+     */
+    fun getBitmapFromUriString(uri: Uri, context: Context, onBitmapLoaded: (Bitmap) -> Unit) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: This isn't getting it from a `Uri String`, but from a `Uri`. Please rename

File: app/src/main/java/com/example/drawingApp/MainActivity.kt [hunk 4]
```diff
@@ -12,6 +14,8 @@ import com.example.drawingApp.CustomViews.DrawingFieldView
 import com.example.drawingApp.DataClasses.Hsv
 import com.example.drawingApp.Utils.ColorPickerUtility
 import com.example.drawingApp.Utils.DialogUtility
+import com.example.drawingApp.Utils.ImageUtility.getBitmapFromUriString
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/MainActivity.kt]: Don't import static methods like this, just call it statically.

File: app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.target.CustomTarget
+import com.bumptech.glide.request.transition.Transition
+
+object ImageUtility {
+
+    /**
+     * launches with the purpose of obtaining an image, will open the users gallery
+     */
+    fun launchGalleryImageGetter(contentLauncher: ActivityResultLauncher<String>) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: This doesn't make sense to be here. It's entirely activity related and should be in the activity

File: app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            .load(uri)
+            .into(BitmapGetter {
+                onBitmapLoaded(it)
+            })
+    }
+}
+
+class BitmapGetter(val onGetBitmap: (Bitmap) -> Unit) : CustomTarget<Bitmap>() {
+    override fun onLoadCleared(placeholder: Drawable?) {}
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: I'd like to see some form of error handling baked in here somewhere. What if the grab doesn't work?

File: app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt (hunk 3 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: renamed

File: app/src/main/java/com/example/drawingApp/MainActivity.kt (hunk 4 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/MainActivity.kt]: fair enough, I believe I should change the import to just stop at the ImageUtility and then I can call them staticaly.

File: app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt (hunk 6 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: implemented some cool ass error handling with glide

Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/Utils/ImageUtility.kt]: also implemented normal boring error handling

File: app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt (hunk 2 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt]: implemented a scope and the ability to not resize in the function

Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/CustomViews/DrawingFieldView.kt]: this View is a customView so unless I extend it as an ImageView I'm not sure if I have the ability to use those ImageView properties

File: app/src/main/java/com/example/drawingApp/DrawingViewModel.kt [hunk 7]
```diff
@@ -112,7 +114,7 @@ class DrawingViewModel : ViewModel() {
     }
 
     fun updateColorFromHsv() {
-        hsv.toColor()?.let {
+        hsv.toColor().let {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/DrawingViewModel.kt:117]: `activeColor = hsv.toColor()`?

File: app/src/main/java/com/example/drawingApp/MainActivity.kt [hunk 8]
```diff
@@ -26,8 +31,21 @@ class MainActivity : AppCompatActivity() {
         enableStrictMode()
         val colorCircleView = findViewById<ColorCircleView>(R.id.colorCircle)
         val drawingFieldView = findViewById<DrawingFieldView>(R.id.drawField)
+        val getGalleryImageView = findViewById<ImageView>(R.id.getImageButton)
         var onSubmission: (String?) -> Boolean
+        val imageContent = registerForActivityResult(ActivityResultContracts.GetContent()) {
+            it?.let { uri ->
+                ImageUtility.getBitmapFromUri(uri, this, {
+                    Toast.makeText(this, "Error getting Bitmap from URI", Toast.LENGTH_SHORT)
+                }) { map ->
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingApp/MainActivity.kt]: map?

Comment (ForzenAcademy) [reply app/src/main/java/com/example/drawingApp/MainActivity.kt]: Rename please

**Review by ForzenAcademy — APPROVED** [review]

File: app/src/main/java/com/example/drawingApp/MainActivity.kt (hunk 8 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/MainActivity.kt]: renamed

File: app/src/main/java/com/example/drawingApp/DrawingViewModel.kt (hunk 7 above)
Comment (HamdanSCode) [reply app/src/main/java/com/example/drawingApp/DrawingViewModel.kt:117]: fixed

**Pull request merged**